research/build-reverse-dependencies/index.md

```diff
@@ -0,0 +1,98 @@
+
+## Mass Prebuild
+
+This is a CLI tool that resolves reverse dependencies and rebuild them all. It can work with both Koji and COPR.
+Even if it is a CLI tool we can not easily integrate it in packit-service because of two main design choices:
+
+- the tool interacts with a database
+- the tool keeps a session and can interact with the user
+
+The above design choices make the command line tool and also the code behind it not reusable as it is.
+
+## Build our own solution
+
+We will need:
+
+1. a way to resolve reverse dependencies
+
+   - we can follow the [Koschei way](https://github.com/fedora-infra/koschei/blob/master/koschei/backend/depsolve.py) which uses `hawkey` from `libdnf`
+   - or we can follow the [Mass prebuild way](https://gitlab.com/fedora/packager-tools/mass-prebuild/-/blob/main/mass_prebuild/utils/whatrequires.py?ref_type=heads): given a package or a set of packages, namely "main packages", the mass pre-builder will calculate the list of its direct reverse dependencies: packages that explicitly mark one of the main packages in their "BuildRequires" field. As far as I can understand it, this is a **less deep** and thus **less complete** approach compared with the Koschei solution.
+   - we can **avoid calculating dependencies** and let the user specify them in a statistical way.
+
+2. a way to schedule builds in Koji or COPR (or both).
+
+   We are quite good at submitting builds in both services but we are not good at dealing with dependencies.
+
+   - **if we don't need to interact with the rebuild process** we can depend upon:
+
+     - the [COPR build batches](https://docs.pagure.org/copr.copr/user_documentation.html#build-batches) feature that allows you to define the order of your builds in advance.
+
+     ```
+           $ copr build <project> --no-wait <first.src.rpm>
+           Created builds: 101010
+           $ copr build <project> --no-wait <second.src.rpm> --after-build-id 101010
+           Created builds: 101020
+           $ copr build <project> --no-wait <third.src.rpm> --with-build-id 101020
+           Created builds: 101030
+     ```
+
+     - or/and the [Koji chained builds](https://docs.fedoraproject.org/en-US/package-maintainers/Using_the_Koji_Build_System/#chained_builds)
+
+     ```
+           fedpkg chain-build --target=side-tag-name libwidget libaselib : libgizmo :
+     ```
+
+   - but we should be aware that **the mass prebuild tool is designed in a way that lets you interact with every single step of the long rebuild process**. I don't know exactly why, but if I got it correctly, this process is error prone and the user may like to be able to continue it from a point of failure or may like to resubmit a failed build. I think it is worth discussing with the mass pre build developer why he has chosen to go this way.
+
+     If we want the user to be able to interact with every step in this long process we should probably be capable of dealing with **job dependencies**.
```

Comment:
I hope the more straightforward way, i.e. using batch builds/ chain builds without the need to solve the dependencies ourselves will be enough. We should probably get feedback from the potential users for this so that we can decide.